msaad.go

```diff
@@ -292,34 +349,54 @@ func (crg *cachedRoleGroups) idToGroupName(i GroupIDU32) string {
 
 // MSAAD is a container for the Microsoft Azure Active Directory synchronization system
 type MSAAD struct {
-	Config ConfigMSAAD
+	msaadConfig ConfigMSAAD
+	provider    MSAADProviderI
+	parent      *Central
+	log         *log.Logger
 
-	parent *Central
+	numAADRoleFetches int
+}
 
-	// Bearer token for communicating with Microsoft Graph API
-	tokenLock      sync.Mutex
-	token          string
-	tokenExpiresAt time.Time
+func (m *MSAAD) Provider() MSAADProviderI {
+	return m.provider
+}
 
-	numAADRoleFetches int
+func (m *MSAAD) SetProvider(provider MSAADProviderI) {
+	m.provider = provider
 }
 
-// Split displayname into firstname, surname
-// "Nick de Jager" -> "Nick" "de Jager"
-// "Abraham Lincoln" -> "Abraham" "Lincoln"
-// "Bad boy Bubby" -> "Bad" "boy Bubby" (this one is wrong, but there's just no way we can tell from a concatenated string)
-func splitDisplayName(dn string) (string, string) {
-	firstSpace := strings.Index(dn, " ")
-	if firstSpace == -1 {
-		return dn, ""
-	}
-	return dn[:firstSpace], dn[firstSpace+1:]
+func (m *MSAAD) Config() ConfigMSAAD {
+	return m.msaadConfig
+}
+
+func (m *MSAAD) Parent() *Central {
+	return m.parent
+}
+
+func (m *MSAAD) SetConfig(msaad ConfigMSAAD) {
+	m.msaadConfig = msaad
 }
 
 // Initialize seeks to initialize the parent context on the MSAAD object
-func (m *MSAAD) Initialize(parent *Central) {
+func (m *MSAAD) Initialize(parent *Central, log *log.Logger) error {
+	if parent == nil {
+		return fmt.Errorf("MSAAD parent parameter is nil")
+	}
 	m.parent = parent
-	m.tokenExpiresAt = time.Now().Add(-time.Hour)
+	if log == nil {
+		return fmt.Errorf("MSAAD logger is nil")
+	}
+
+	m.log = log
+	msaadProvider := MSAADProvider{}
+	err := msaadProvider.Initialize(m, m.log)
+	if err != nil {
+		return fmt.Errorf("could not initialise MSAAD provider")
+	}
+	msaadProvider.tokenExpiresAt = time.Now().Add(-time.Hour)
```

Comment:
This is pre-existing code from initialisation of MSAAD, and since it is effectively a global on the MSAAD itself, it was probably just a safeguard to not have a default value that is actually valid. Possibly came out during testin years ago and was just set to be in the past by default:

```
func (m *MSAAD) Initialize(parent *Central) {
	m.parent = parent
	m.tokenExpiresAt = time.Now().Add(-time.Hour)
}
```